Use protego as a default robots.txt parser (#4006)
scrapy/scrapy: docs/topics/downloader-middleware.rst

the request will be ignored by this middleware even if
:setting:`ROBOTSTXT_OBEY` is enabled.

Parsers vary in several aspects:

* Language of implementation

* Supported specification

* Support for wildcard matching

> * Usage of length based rule: in particular for ``Allow`` and

Reviewer: :lipstick: I wonder if we should link to some external resource here for users who want to know more about this, e.g. https://developers.google.com/search/reference/robots_txt#order-of-precedence-for-group-member-lines

  ``Disallow`` directives, where the most specific rule based on the length of
  the path trumps the less specific (shorter) rule

.. _protego-parser:

Protego parser
~~~~~~~~~~~~~~

Based on `Protego <https://github.com/scrapy/protego>`_:

* implemented in Python

* is compliant with `Google's Robots.txt Specification
  <https://developers.google.com/search/reference/robots_txt>`_

* supports wildcard matching

* uses the length based rule

Scrapy uses this parser by default.

.. _python-robotfileparser:

RobotFileParser
~~~~~~~~~~~~~~~

Based on `RobotFileParser
<https://docs.python.org/3.7/library/urllib.robotparser.html>`_:

> * is Python's built-in robots.txt_ parser.

Reviewer: :lipstick: List entries of parsers are inconsistent in whether or not they end in a period. It would be great if they all ended the same way. Since they are not capitalized, I would remove the period from the ones that currently have one.

* is compliant with `Martijn Koster's 1996 draft specification
  <http://www.robotstxt.org/norobots-rfc.txt>`_.

* lacks support for wildcard matching.

* doesn't use the length based rule

It is faster than Protego and backward-compatible with versions of Scrapy before 1.8.0 .

In order to use this parser, set:

* :setting:`ROBOTSTXT_PARSER` to ``scrapy.robotstxt.PythonRobotParser``

.. _reppy-parser:

Reppy parser
~~~~~~~~~~~~

Based on `Reppy <https://github.com/seomoz/reppy/>`_:

* is a Python wrapper around `Robots Exclusion Protocol Parser for C++
  <https://github.com/seomoz/rep-cpp>`_.

* is compliant with `Martijn Koster's 1996 draft specification
  <http://www.robotstxt.org/norobots-rfc.txt>`_.

* supports wildcard matching

* uses the length based rule

> Native implementation provides better speed than Protego.

Reviewer: Should we indicate, in average, how the performance of different parsers compares to the default parser?

* Usage of `length based rule <https://developers.google.com/search/reference/robots_txt#order-of-precedence-for-group-member-lines>`_:
  in particular for ``Allow`` and ``Disallow`` directives, where the most
  specific rule based on the length of the path trumps the less specific
  (shorter) rule

Performance comparison of different parsers is available at `the following link
<https://anubhavp28.github.io/gsoc-weekly-checkin-12/>`_.

.. _protego-parser:

Protego parser
~~~~~~~~~~~~~~

Based on `Protego <https://github.com/scrapy/protego>`_:

* implemented in Python

* is compliant with `Google's Robots.txt Specification
  <https://developers.google.com/search/reference/robots_txt>`_

* supports wildcard matching

* uses the length based rule

Scrapy uses this parser by default.

.. _python-robotfileparser:

RobotFileParser
~~~~~~~~~~~~~~~

Based on `RobotFileParser
<https://docs.python.org/3.7/library/urllib.robotparser.html>`_:

* is Python's built-in robots.txt_ parser

* is compliant with `Martijn Koster's 1996 draft specification
  <http://www.robotstxt.org/norobots-rfc.txt>`_

* lacks support for wildcard matching

* doesn't use the length based rule

> It is faster than Protego and backward-compatible with versions of Scrapy before 1.8.0 .

Reviewer: :lipstick: 

```suggestion
It is faster than Protego and backward-compatible with versions of Scrapy before 1.8.0.
```

In order to use this parser, set:

* :setting:`ROBOTSTXT_PARSER` to ``scrapy.robotstxt.PythonRobotParser``

.. _reppy-parser:

Reppy parser
~~~~~~~~~~~~

Based on `Reppy <https://github.com/seomoz/reppy/>`_:

* is a Python wrapper around `Robots Exclusion Protocol Parser for C++
  <https://github.com/seomoz/rep-cpp>`_

* is compliant with `Martijn Koster's 1996 draft specification
  <http://www.robotstxt.org/norobots-rfc.txt>`_

* supports wildcard matching

* uses the length based rule

> Native implementation provides better speed than Protego.

Reviewer: :lipstick: 

```suggestion
Native implementation, provides better speed than Protego.
```

RobotFileParser
~~~~~~~~~~~~~~~

`RobotFileParser <https://docs.python.org/3.7/library/urllib.robotparser.html>`_ is

> Python's inbuilt robots.txt_ parser. The parser is fully compliant with `Martijn Koster's

Reviewer 2: ```suggestion
Python's built-in robots.txt_ parser. The parser is fully compliant with `Martijn Koster's
```
I know this line was like this before, but since it's being changed we might as well fix it.

1996 draft specification <http://www.robotstxt.org/norobots-rfc.txt>`_. It lacks

> support for wildcard matching.

Reviewer 3: Probably makes sense to add a note on why this parser is useful:
* faster than protego
* compatibility with previous Scrapy versions

Reviewer 3: RobotFileParser is still mentioned as default in docs above, where all parsers are listed. Could you please re-read the whole docs about robots.txt, and check that they make sense, assuming protego is a default parser?

`Protego <https://github.com/scrapy/protego>`_ is a pure-Python robots.txt_ parser.
The parser is fully compliant with `Google's Robots.txt Specification
<https://developers.google.com/search/reference/robots_txt>`_ hence supports wildcard
matching, and uses the length based rule similar to `Reppy <https://github.com/seomoz/reppy/>`_.

> Scrapy uses this parser by default.

Reviewer 3: It could make sense to change the order of the introduction of different parsers, so that the default one goes first.

Reviewer 3: If we move it to the top, description of this parser shouldn't mention reppy, it would be confusing otherwise.

In order to use this parser:

> 

Reviewer 3: currently you still need to install protego manually; this issue is hidden on CI because protego is included in all requirements.txt files. To fix it, protego needs to be added to setup.py install_requires.
